File: backends/ts-hono/src/usecase/invitation/interactor.ts
/**
 * 招待ユースケース Interactor モジュール。
 *
 * @author Satoshi Nagashiba <satoshi.nagashiba@gmail.com>
 */
import { randomBytes } from "crypto";
import { notFound } from "../../lib/errors.js";
import type { InvitationRepository } from "../../domain/invitation/repository.js";
import type { InvitationAuthRepository } from "../../domain/invitation/authRepository.js";
import type { InvitationResult } from "../../domain/invitation/valueObjects.js";
import { config } from "../../config.js";

function buildResult(token: string): InvitationResult {
  const url = `${config.app.frontendUrl}/invitation/${token}`;
  return { token, url, displayUrl: url.replace(/^https?:\/\//, "") };
}

/** 招待のユースケース実装。 */
export class InvitationInteractor {
  constructor(
    private readonly repo: InvitationRepository,
    private readonly authRepo: InvitationAuthRepository,
  ) {}

  /**
   * 最新の招待情報の VO を返します。
   * @param role - ロール (1=admin, 2=member)
   * @returns InvitationResult
   * @throws AppError 招待が存在しない場合
   */
  async current(role: number): Promise<InvitationResult> {
    const inv = await this.repo.getCurrentByRole(role);
    if (!inv) throw notFound("invitation_not_found");
    return buildResult(inv.token);
  }

  /**
   * 新しい招待トークンを発行し、VO を返します。
   * @param role - ロール (1=admin, 2=member)
   * @returns InvitationResult
   */
  async issue(role: number): Promise<InvitationResult> {
    const token = randomBytes(16).toString("hex");
    const inv = await this.repo.issue(token, role);
    return buildResult(inv.token);
  }

  /**
   * トークンで招待情報の VO を返します。
   * @param token - 招待トークン
   * @returns InvitationResult
   * @throws AppError 招待が存在しない場合
   */
  async findByToken(token: string): Promise<InvitationResult> {
    const inv = await this.repo.findByToken(token);
    if (!inv) throw notFound("invitation_not_found");
    await this.authRepo.store(inv.token, inv.role, 600);
    return buildResult(inv.token);
  }
}

File: backends/python-fastapi/app/usecase/invitation/interactor.py
"""
招待ユースケース Interactor モジュール。

Author: Satoshi Nagashiba <satoshi.nagashiba@gmail.com>
"""
from app.domain.invitation.value_objects import InvitationVo
from app.domain.invitation.repository import InvitationRepository
from app.domain.invitation.auth_repository import InvitationAuthRepository
from app.exceptions import not_found


class InvitationInteractor:
    """招待のユースケース実装。

    Attributes:
        invitation_repository: 招待リポジトリ
        invitation_auth_repository: 招待認証キャッシュリポジトリ
    """

    def __init__(self, repo: InvitationRepository, auth_repo: InvitationAuthRepository):
        """初期化します。

        Args:
            repo: 招待リポジトリ
            auth_repo: 招待認証キャッシュリポジトリ
        """
        self.invitation_repository = repo
        self.invitation_auth_repository = auth_repo

    def current(self, role: int) -> InvitationVo:
        """ロールで絞り込んだ最新の招待情報の Vo を返します。

        Args:
            role: ロール（1=管理者, 2=メンバー）

        Returns:
            InvitationVo インスタンス

        Raises:
            AppException: 招待が存在しない場合
        """
        result = self.invitation_repository.get_current_by_role(role)
        if result is None:
            raise not_found("invitation_not_found")
        return result

    def issue(self, role: int) -> InvitationVo:
        """新しい招待トークンを発行し、招待情報の Vo を返します。

        Args:
            role: ロール（1=管理者, 2=メンバー）

        Returns:
            InvitationVo インスタンス
        """
        return self.invitation_repository.issue(role)

    def find_by_token(self, token: str) -> InvitationVo:
        """トークンで招待情報の Vo を返し、認証トークンとロールをキャッシュします。

        Args:
            token: 招待トークン

        Returns:
            InvitationVo インスタンス

        Raises:
            AppException: 招待が存在しない場合
        """
        result = self.invitation_repository.find_by_token(token)
        if result is None:
            raise not_found("invitation_not_found")
        self.invitation_auth_repository.store(result.token, result.role, 600)
        return result
